File: v0/lib/vector2d.dart
class Vector2d {
  double x, y;

  Vector2d(this.x, this.y);

  operator +(Vector2d v) => Vector2d(x + v.x, y + v.y);
  operator -(Vector2d v) => Vector2d(x - v.x, y - v.y);
  operator *(Vector2d v) => Vector2d(x * v.x, y * v.y);

  Vector2d copy() => Vector2d(x, y);
  Vector2d addN(double n) {
    x += n;
    y += n;
    return this;
  }

  Vector2d subtractN(double n) {
    x -= n;
    y -= n;
    return this;
  }

  Vector2d multiplyN(double n) {
    x *= n;
    y *= n;
    return this;
  }

  Vector2d divideN(double n) {
    if (n == 0) throw Exception("cannot divide by zero");
    x /= n;
    y /= n;
    return this;
  }
}

File: physics-playground/src/Vector2d.ts
export class Vector2d {
    public constructor(public x: number, public y: number) {}

    public copy(): Vector2d {
        return new Vector2d(this.x, this.y);
    }

    public add({ x, y }: Vector2d): Vector2d {
        this.x += x;
        this.y += y;
        return this;
    }

    public subract({ x, y }: Vector2d): Vector2d {
        this.x -= x;
        this.y -= y;
        return this;
    }

    public multiply({ x, y }: Vector2d): Vector2d {
        this.x *= x;
        this.y *= y;
        return this;
    }

    public addN(v: number): Vector2d {
        this.x += v;
        this.y += v;
        return this;
    }

    public subractN(v: number): Vector2d {
        this.x -= v;
        this.y -= v;
        return this;
    }

    public multiplyN(v: number): Vector2d {
        this.x *= v;
        this.y *= v;
        return this;
    }

    public divideN(v: number): Vector2d {
        if (v === 0) throw new Error("cannot divide by zero");
        this.x /= v;
        this.y /= v;
        return this;
    }
}

export const vec2d = (x = 0, y = x) => new Vector2d(x, y);
